`for ... in obj` where `obj: object` gives TS2407

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  2.2.0

```
$ node_modules/.bin/tsc --version
Version 2.2.0
```

**Code**

```ts
let obj: object = {}

for (const attr in obj) {
  const val = obj[attr]
}
```

**Expected behavior:**

Compiles without errors.

**Actual behavior:**

```
some.ts(3,20): error TS2407: The right-hand side of a 'for...in' statement must be of type 'any', an object type or a type parameter.
```

This is a new feature, so I'm not sure if I'm using this properly, but the error message says that `an object type` is allowed, so I expect this behavior to be a bug.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1,4 +1,4 @@
-﻿/// <reference path="moduleNameResolver.ts"/>
+/// <reference path="moduleNameResolver.ts"/>
 /// <reference path="binder.ts"/>
 
 /* @internal */
@@ -18606,7 +18606,7 @@ namespace ts {
 
             // unknownType is returned i.e. if node.expression is identifier whose name cannot be resolved
             // in this case error about missing name is already reported - do not report extra one
-            if (!isTypeAnyOrAllConstituentTypesHaveKind(rightType, TypeFlags.Object | TypeFlags.TypeVariable)) {
+            if (!isTypeAnyOrAllConstituentTypesHaveKind(rightType, TypeFlags.Object | TypeFlags.TypeVariable | TypeFlags.NonPrimitive)) {
                 error(node.expression, Diagnostics.The_right_hand_side_of_a_for_in_statement_must_be_of_type_any_an_object_type_or_a_type_parameter);
             }
 

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForIn.js
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForIn.js
@@ -1,0 +1,13 @@
+//// [nonPrimitiveIndexingWithForIn.ts]
+var a: object;
+
+for (var key in a) {
+    var value = a[key];
+}
+
+
+//// [nonPrimitiveIndexingWithForIn.js]
+var a;
+for (var key in a) {
+    var value = a[key];
+}

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForIn.symbols
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForIn.symbols
@@ -1,0 +1,14 @@
+=== tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForIn.ts ===
+var a: object;
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForIn.ts, 0, 3))
+
+for (var key in a) {
+>key : Symbol(key, Decl(nonPrimitiveIndexingWithForIn.ts, 2, 8))
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForIn.ts, 0, 3))
+
+    var value = a[key];
+>value : Symbol(value, Decl(nonPrimitiveIndexingWithForIn.ts, 3, 7))
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForIn.ts, 0, 3))
+>key : Symbol(key, Decl(nonPrimitiveIndexingWithForIn.ts, 2, 8))
+}
+

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForIn.types
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForIn.types
@@ -1,0 +1,15 @@
+=== tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForIn.ts ===
+var a: object;
+>a : object
+
+for (var key in a) {
+>key : string
+>a : object
+
+    var value = a[key];
+>value : any
+>a[key] : any
+>a : object
+>key : string
+}
+

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForInNoImplicitAny.errors.txt
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForInNoImplicitAny.errors.txt
@@ -1,0 +1,12 @@
+tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInNoImplicitAny.ts(4,17): error TS7017: Element implicitly has an 'any' type because type '{}' has no index signature.
+
+
+==== tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInNoImplicitAny.ts (1 errors) ====
+    var a: object;
+    
+    for (var key in a) {
+        var value = a[key]; // error
+                    ~~~~~~
+!!! error TS7017: Element implicitly has an 'any' type because type '{}' has no index signature.
+    }
+    

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForInNoImplicitAny.js
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForInNoImplicitAny.js
@@ -1,0 +1,13 @@
+//// [nonPrimitiveIndexingWithForInNoImplicitAny.ts]
+var a: object;
+
+for (var key in a) {
+    var value = a[key]; // error
+}
+
+
+//// [nonPrimitiveIndexingWithForInNoImplicitAny.js]
+var a;
+for (var key in a) {
+    var value = a[key]; // error
+}

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.js
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.js
@@ -1,0 +1,13 @@
+//// [nonPrimitiveIndexingWithForInSupressError.ts]
+var a: object;
+
+for (var key in a) {
+    var value = a[key];
+}
+
+
+//// [nonPrimitiveIndexingWithForInSupressError.js]
+var a;
+for (var key in a) {
+    var value = a[key];
+}

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.symbols
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.symbols
@@ -1,0 +1,14 @@
+=== tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInSupressError.ts ===
+var a: object;
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 0, 3))
+
+for (var key in a) {
+>key : Symbol(key, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 2, 8))
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 0, 3))
+
+    var value = a[key];
+>value : Symbol(value, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 3, 7))
+>a : Symbol(a, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 0, 3))
+>key : Symbol(key, Decl(nonPrimitiveIndexingWithForInSupressError.ts, 2, 8))
+}
+

--- a/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.types
+++ b/tests/baselines/reference/nonPrimitiveIndexingWithForInSupressError.types
@@ -1,0 +1,15 @@
+=== tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInSupressError.ts ===
+var a: object;
+>a : object
+
+for (var key in a) {
+>key : string
+>a : object
+
+    var value = a[key];
+>value : any
+>a[key] : any
+>a : object
+>key : string
+}
+

--- a/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForIn.ts
+++ b/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForIn.ts
@@ -1,0 +1,5 @@
+var a: object;
+
+for (var key in a) {
+    var value = a[key];
+}

--- a/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInNoImplicitAny.ts
+++ b/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInNoImplicitAny.ts
@@ -1,0 +1,6 @@
+// @noImplicitAny: true
+var a: object;
+
+for (var key in a) {
+    var value = a[key]; // error
+}

--- a/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInSupressError.ts
+++ b/tests/cases/conformance/types/nonPrimitive/nonPrimitiveIndexingWithForInSupressError.ts
@@ -1,0 +1,7 @@
+// @noImplicitAny: true
+// @suppressImplicitAnyIndexErrors: true
+var a: object;
+
+for (var key in a) {
+    var value = a[key];
+}

